lfs/gitscanner_log.go

```diff
@@ -63,6 +64,32 @@ func scanUnpushed(cb GitScannerFoundPointer, remote string) error {
 	return nil
 }
 
+func scanStashed(cb GitScannerFoundPointer, s *GitScanner) error {
+	// First get the SHAs of all stashes
+	// git reflog show --format="%H" stash
+	reflogArgs := []string{"show", "--format=%H", "stash"}
+
+	cmd, err := git.RefLog(reflogArgs...)
+	if err != nil {
+		return err
+	}
+
+	cmd.Start()
+	defer cmd.Wait()
+
+	scanner := bufio.NewScanner(cmd.Stdout)
+
+	for scanner.Scan() {
+		err = s.ScanRef(strings.TrimSpace(scanner.Text()), cb)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
```

Comment:
After a bit of a deeper look, I had a few notes on this `scanStashed()` function.

The first thing which puzzled me for a while was that it seemed to succeed in a freshly initialized, empty repo with no stashes, where `.git/refs/stash` doesn't exist yet.  I couldn't get `git reflog show stash` to work on the command line, but it seemed to be OK when run this way.  But the immediate issue is that the possible error exit code returned by `cmd.Wait()` is not being checked.  If you add to the end of this function something like:
```go
        stderr, _ := ioutil.ReadAll(cmd.Stderr)
        err = cmd.Wait()
        if err != nil {
                return fmt.Errorf("error in git reflog: %v %v", err, string(stderr))
        }
```

then it does fail in a repo where there's no `refs/stash` ref, or if you change `"stash"` to `"foobar"` or something.  So that's the first issue, that the Git command's exit code is not checked.

Looking at how the `git stash list` command works, I believe it starts by [checking](https://github.com/git/git/blob/dc04167d378fb29d30e1647ff6ff51dd182bc9a3/builtin/stash.c#L691-L692) for an extant `refs/stash` ref, and exiting early if it doesn't exist.

So we probably need something similar here once we're checking the Git command's exit code; otherwise we'll bail with `Prune error: error in git reflog: exit status 128 fatal: ambiguous argument 'stash': unknown revision or path not in the working tree.`  (That's what I get once I added the check above on `cmd.Wait()`'s return value, plus failures in the `t/t-prune.sh` and other tests.)

Beyond that, I looking again at how `git stash list` [uses](https://github.com/git/git/blob/dc04167d378fb29d30e1647ff6ff51dd182bc9a3/builtin/stash.c#L695-L696) the `git log -g refs/stash --` command, I wondered if we could avoid the need for a new `git.RefLog()` function in `git/git.go` entirely and just use `git.Log()`?

Reply:
I worked up a tentative first half of an alternate version of this function, which checks for `refs/stash` first, and then uses `git.Log()`.  The hard part, which I skipped for now, is reading the output; I thought that could borrow from the design of `parseScannerLogOutput()` but without the need for a `direction` parameter.

```go
func scanStashed(cb GitScannerFoundPointer, s *GitScanner) error {
        if _, err := git.ResolveRef("refs/stash"); err != nil {
                return nil
        }

        logArgs := []string{"-g"}

        // Add standard search args to find lfs references
        logArgs = append(logArgs, logLfsSearchArgs...)
        // Search reflog (with -g) for refs/stash
        logArgs = append(logArgs, "refs/stash", "--")

        cmd, err := git.Log(logArgs...)
        if err != nil {
                return err
        }
```

Reply:
Hmm, personally, I think the original version is more elegant - I know it's not checking the result code, but this is a pattern that already exists in the codebase (see RemoteList, RemoteURLs, RecentBranches). The only reason for this command to fail (and others to not) is that there are no stashes to list.

I didn't use `git log` because there's no need to "walk" the history - all stashes are singular commits. reflog is more efficient to get the data needed and I thought the RefLog command might be useful later on. But I don't mind if you'd prefer to write it another way.

Reply:
Ah, I see the similarity to `RemoteList` and friends now; I had not thought to look at those, only at the other functions nearby.  Well, given that, maybe there's a tidy way to combine the two approaches and get a minimal set of changes?

Here's what seems to be working for me tonight, replacing just the `git reflog show` with `git log -g`, i.e., something like:
```go
	// git log -g --format="%H" refs/stash --
	logArgs := []string{"-g", "--format=%H", "refs/stash", "--"}

	cmd, err := git.Log(logArgs...)
```

I can't claim to be an expert on the exact functional differences between those two commands (but I'll ask an expert :-) ... however, if they're effectively the same for this purpose, that would allow for reuse of the `git.Log()` function and avoid the need to add `git.RefLog()`.  Although I confess this is a very small thing, really; I could go either way (particularly if it turns out there's a specific advantage to the `git reflog` command)!

Thanks again so much for working on this; it's very much appreciated.

Reply:
OK you're right I think `-g` is basically an alias for reflog and makes log behave very differently. Unfortunately for the same reason `-g` makes some of the other options in `logLfsSearchArgs` to `git log` stop working, such as the `-G "oid sha256:" search, so we can't combine everything into one `git log` and send that  to `parseScannerLogOutput`. So you're still down to executing 2 commands, albeit that both of them are now `git log` so the RefLog command isn't needed.

Reply:
Yeah, I tossed out my attempt to use `logLfsSearchArgs`; that was a non-starter.  All I was thinking of was changing the first couple of lines of this function to use `git.Log()` and the `git log -g ...` command format.  (And according to the folks I consulted, `git log -g` is equivalent to `git reflog show`.

That said, based on the feedback I got, I realized there may be a different, more "interesting" issue yet to solve here, which is what happens when someone uses `git stash pop --index` and had LFS file changes in both their index *and* their working tree.  I'll write that up in the main comments, though.